Efficient grid search article (#244)
tidymodels/tidymodels.org: learn/work/efficient-grid-search/index.qmd

When people talk about grid search, they tend to focus on all possible combinations of our candidate parameter values (we'll discuss other grids below). In our case, we have a $2\times 3 = 6$ candidate grid. In tidymodels:

```{r}
#| label: make-grid-a
library(tidymodels)
parameters(neighbors(c(1, 30)), min_n(c(1, 100))) |>
  grid_regular(levels = c(2, 3)) |>
  arrange(neighbors, min_n)
```

Visually:

```{r}
#| label: grid-a
#| echo: false
#| out-width: 50%
#| fig-width: 4
#| fig-height: 4
#| fig-align: center
parameters(neighbors(c(1, 30)), min_n(c(1, 100))) |>
  grid_regular(levels = c(2, 3)) |>
  arrange(neighbors, min_n) |> 
  ggplot(aes(neighbors, min_n)) + 
  geom_point() + 
  labs(x = "# Neighbors", y = "Min N")
```

To tune our model, we would iterate over this grid, fitting six different models and evaluating their performance to choose the best one.

There's a problem, though: UMAP is a very expensive preprocessing method. If we loop through the rows of our grid, we'll end up estimating UMAP times even though there are only two possible values. That's incredibly inefficient.

> ### Problem: DRY (Don't Repeat Yourself)

Reviewer: The negation here is potentially confusing/misleading. Repetition is the problem, not the "don't"/avoidance thereof.

If we have an inexpensive preprocessor for our data, repeating it is not a big problem; we would probably just keep things simple and effectively use a `for` loop over our rows.

However, there is a much better solution to make the computations as inexpensive as possible.

### Solution: Stagewise Estimation and Conditional Execution

> We can separate our hyperparameters into three possible sets for the three possible stages of the pipeline: preprocessing, the supervised model, and postprocessing. In our example, we have a single tuning parameter for preprocessing and the supervised model.

Reviewer: ```suggestion
We can separate our hyperparameters into three possible sets for the three possible stages of the pipeline: preprocessing, the supervised model, and postprocessing. In our example, we have a single tuning parameter each for preprocessing and the supervised model.
```

    # Remove axis lines
    axis.line = element_blank(),
    # Remove axis ticks
    axis.ticks = element_blank(),
    # Remove the border around the plot panel
    panel.border = element_blank()
  )  +
  labs(y = NULL, title = "1 Neighbor, Min N 50")

iter_3 <- 
  cond_1 |>
  ggplot(aes(x = parameter, y = configuration))  +
  geom_tile(aes(fill = col_val_c), show.legend = FALSE) +
  scale_fill_identity() +
  scale_x_discrete(position = "top") +
  theme(
    # Remove background grid (reference lines)
    panel.grid = element_blank(),
    # Remove axis lines
    axis.line = element_blank(),
    # Remove axis ticks
    axis.ticks = element_blank(),
    # Remove the border around the plot panel
    panel.border = element_blank()
  )  +
  labs(y = NULL, title = "1 Neighbor, Min N 100")
iter_1 + iter_2 + iter_3
```

We end up training six different tree ensembles but estimate only two UMAP models. This strategy can save a lot of time, especially for regular grids.

This conditional execution strategy seems like a simple solution, and it is. However, it will become more complex if we recursively iterate over the model and post-processing parameters.

## Problem: Curse of Dimensionality

> Part of the problem with the conventional wisdom that "grid search is massively inefficient" is the Curse of Dimensionality. As the problem dimensions increase, the volume of the search space becomes vast. In our case, this means that, as we add more tuning parameters, the parameter space explodes exponentially. For example, if we wanted to evaluate four values of each of $p$ tuning parameters, we would need $p^d$ combinations. With four parameters that are evaluated on five distinct points, the grid would have 1,024 rows.

Reviewer: ```suggestion
Part of the problem with the conventional wisdom that "grid search is massively inefficient" is the Curse of Dimensionality. As the problem dimensions increase, the volume of the search space becomes vast. In our case, this means that, as we add more tuning parameters, the parameter space explodes exponentially. For example, if we wanted to evaluate $d$ values of each of $p$ tuning parameters, we would need $d^p$ combinations. With five parameters that are evaluated on four distinct points, the grid would have 1,024 rows.
```

It's a difficult situation as long as you don't think about the problem very hard. There are two ways to avoid this issue:

> - [Parallelize](http://tmwr.org/grid-search#parallel-processing) the tasks. If I can do 10 things at once, the cost of doing more drops significantly. Grid search is an "embarrassingly parallel" problem, and running the computations in parallel can produce speedups of several-fold. This is not hard to do in R or tidymodels.

Reviewer: everything else is "we"

```{r}
#| label: mixed
#| echo: false
#| out-width: 70%
#| fig-width: 5
#| fig-height: 5
#| fig-align: center
umap_sub_param <- parameters(neighbors(c(1, 30)), min_n(c(1, 100)))

sub_grid <- umap_sub_param |> grid_space_filling(size = 30, type = "uniform")

mixed_grid <- sub_grid |> crossing(trees = value_seq(trees(c(1, 100)), 30))

mixed_grid|>
  ggplot(aes(x = .panel_x, y = .panel_y)) +
  geom_point(cex = 1 / 2) +
  facet_matrix(vars(everything()), layer.diag = 2, grid.y.diag = FALSE)
```

Recall that this is only really an issue when our model has submodel parameters and is inefficient to train. As a counterexample, the aforementioned glmnet model fits generalized linear models (e.g., linear or logistic regression) but includes a penalty parameter that can reduce model complexity, lowering the risk of overfitting. From a single model, it can quickly produce predictions for a large number of predictions across a sequence of penalty values. As with the tree example, an SFD would negate the submodel benefit, but the glmnet model trains very quickly, so an SFD would still be effective for tuning.

## Postprocessing

> Postprocessing methods transform predictions to make them more useful. We can think of them as being estimated or not. As was mentioed in the first section, if we have a classification model with two possible classes, we might want to adjust the rule that converts the predicted class probabilities to "event" or "no event". This enables us to optimize for what we care about (e.g., sensitivity or specificity). For this example, we are not estimating a parameter to test a potential threshold; we simply apply something like `ifelse(prob >= threshold, "event", "no event")` to transform the class predictions.

Reviewer: ```suggestion
Postprocessing methods transform predictions to make them more useful. We can think of them as being either estimated or not. As was mentioned in the first section, if we have a classification model with two possible classes, we might want to adjust the rule that converts the predicted class probabilities to "event" or "no event". This enables us to optimize for what we care about (e.g., sensitivity or specificity). For this example, we are not estimating a parameter to test a potential threshold; we simply apply something like `ifelse(prob >= threshold, "event", "no event")` to transform the class predictions.
```
